### ISSUE TITLE
Can not format via ssh: There is no formatter for 'cpp'-files installed

### ISSUE THREAD

**Issue by mohsenomidi** [microsoft/vscode-cpptools#3787]
Hi,

C++ formatter on remote host 

**Describe the bug**
- OS and Version: Microsoft Windows 10 Enterprise (10.0.17763 N/A Build 17763)
- VS Code Version: 1.35.1
- C/C++ Extension Version: 023.1
- Other extensions you installed (and if the issue persists after disabling them): "Kelvin Schoofs(SSH FS) version (1.16.2)
- A clear and concise description of what the bug is:

I am using vscode for my C++ project, i installed the C++ extension (0.23.1) i works fine when i open the file from my local drive.

But i am using "Kelvin Schoofs(SSH FS) version (1.16.2) extension for my remote code on the Linux machine, after opening the C++ file from sshfs work-space, and trying to format my C++ document, it returns "There is no formatter for 'cpp'-files installed."

how can i fix this behavior for remote code "i mentioned again: formatting works properly for local files"

**To Reproduce**
Open the VSCODE and explore your files via the ssh (remote host)
install the "Kelvin Schoofs(SSH FS) and configure it to connect to the remote Linux host. 
the open your files and try to format the document.

**Expected behavior**

Format the C++ Document.

Regards.

**Comment by sean-mcmanus** [comment]
It sounds like you need to install the C/C++ extension or some other formatting extension on the remote machine. Are you able to format with some other extension like the clang-format extension? That "there is no formatter" message is from VS Code itself, so I'm not sure how our extension could be causing it.

**Comment by mohsenomidi** [comment]
@sean-mcmanus thanks for your response. 
for more clarification, my remote host is Ubuntu Linux Box, and it just like the remote drive, there is no development tools on it, and it can only accessible via ssh, then i can access on those files via ssh, therefore it is impossible to use any extension on that machine to format the codes. 

Now i will access to my Linux files via network(SSH) and i edit and manipulate the files with VS Code IDE, VS Code can not access remote Linux machine by itself, therefore i install the extension to be able to connect to remote machine via network (Kelvin Schoofs(SSH FS)), the C/C++ extension can not format my cpp files via the network (SSH), this behavior is the same on other formatting extensions (clang-format is the same), but these extensions work properly on local files.

i don't know the VS Code architecture and nor these extensions, i thought if the file opens by IDE, the extension have full access to the document and if a user request to format document, the extension will be read and cache the file and made the changes, and pass it to IDE, now i don't know the procedure for reporting it, if it related to VS Code, let me know to report it to VS Code Git page, or may be it need more clarification from your side as a collaborator to describe the exact problem, based on your strong knowledge in these system architecture. 

PS: i also opened an issue on clang-format Git, The issue : [ISSUE : #89](https://github.com/xaverh/vscode-clang-format-provider/issues/89)

**Comment by sean-mcmanus** [comment]
Hmm, it may also be a bug with the Kelvin Schoofs(SSH FS) extension, i.e. if the bug also repros with the clang-format extension. You could try using an alternative SSH extension like "Remote - SSH".

**Comment by mohsenomidi** [comment]
@sean-mcmanus: i am trying to configure Microsoft/ Remote SSH, it is too hard to configure the extension, but in my first try, i can not use this extension to connect. 
if i have any success to use this extension for connecting will make comment

**Comment by dbsxdbsx** [comment]
@sean-mcmanus ,Hi, I still got  this issue with latest vscode and remote-SSH. Isn't it is fixed according to some posts from you commented in other issues.  Or should we still install vscode in remote machine?

**Comment by sean-mcmanus** [comment]
@dbsxdbsx I don't believe this is fixed...I'm not sure what issue you're referring to. We have issue https://github.com/microsoft/vscode-cpptools/issues/4121 for correctly handling the clang_format_path setting in the remote scenario.

**Comment by dbsxdbsx** [comment]
@sean-mcmanus ,sorry for some wrong info. And at present, I need to install LLVM to format( with clang-format installed) on remote machine, but no need to install vscode there. I think it is ok , but if LLVM is no need to install either, then it would be perfect.

**Comment by sean-mcmanus** [comment]
@bobbrow Why was this issue marked fixed? I'm not aware of a fix. Has anyone tried the repro?

**Comment by bobbrow** [comment]
@sean-mcmanus We misread the issue and thought it was about standard ssh.  This still needs to be investigated.

**Comment by jakobkilian** [comment]
> @sean-mcmanus We misread the issue and thought it was about standard ssh. This still needs to be investigated.

 I am using standard-ssh with vscode on mac OSX 10.13 and I am actually running in the same issue described: "There is no formatter for 'cpp'-files installed"

**Comment by efreibe** [comment]
Hello, I have installed Remote Workspace for vscode (via ssh) and the same happens for me.

If I try to format a local javascript file, it works, but when I load a remote file it shows: "There is no formatter for 'javascript' files installed."

I believe that the code is loaded into memory so I don't know why it refuses to format it, I've tried with different format providers, so I think that the problem is from vscode itself.

https://github.com/mkloubert/vscode-remote-workspace/issues/2

**Comment by MaxmeTrn** [comment]
I currently have the same issue while using VSCode Windows with the extension SSHFS and the server is under Linux distribution.

**Comment by lijgame** [comment]
I have same issues when I remote to Ubuntu machine from Mac OS.

Everything works fine if I use VSCode locally on Ubuntu machine.
Seems a bug in Remote SSH?

**Comment by jayagami** [comment]
Same, unable to format `cpp `and `cpp-cuda`,  `clangd` must be installed
